### ISSUE TITLE
Remove usage of from_legacy_cache

### ISSUE THREAD

**Issue by BenjaminBossan** [huggingface/peft#2757]
We use the `from_legacy_cache` method of transformers `Cache` classes in PEFT. This method will be removed from transformers in v4.58.0. PEFT should remove this code and find a different solution before then.

**Comment by SP1029** [comment]
Greetings @BenjaminBossan 

I tried going over this, and the following was what I was able to understand. Could you please provide suggestions?

**Existing Situation**

`from_legacy_cache` is being used in `peft` in the following, where it is being primarily used to convert key-values to a `cache` object for the `DynamicCache` and the `EncoderDecoderCache`.

https://github.com/huggingface/peft/blob/a3197b1ec54baa107e2efc4a30c19209d9ea489a/src/peft/peft_model.py#L789-L796


In the `transformers` repo, the following is the main file for our purposes:

https://github.com/huggingface/transformers/blob/1363fceeec955db125141c3d7768f8ef345ed288/src/transformers/cache_utils.py

The overall structure is as follows. `Cache` is an object consisting of one cache per layer, where each layer's cache stores the keys and corresponding values in two tensors. `DynamicCache` is a variant of `Cache`, allowing updates to the keys and values by appending new keys and values on calling `update`. `EncoderDecoderCache` consists of two `DynamicCache` objects. `from_legacy_cache` converts the cache in the form of tuples to these cache objects.

**Possible Solution**

For `DynamicCache` in `init`, there is an argument `ddp_cache_data` which takes the data of the cache, the `past_key_values` can be passed as this argument, since `from_legacy_cache` also seems to process in a similar manner. Similarly, for the `caches` in the `init` function of `EncoderDecoderCache`.
```diff
- 789 past_key_values = DynamicCache.from_legacy_cache(past_key_values)
+ 789 past_key_values = DynamicCache(past_key_values)
- 796 past_key_values = EncoderDecoderCache.from_legacy_cache(past_key_values)
+ 796 past_key_values = EncoderDecoderCache(past_key_values)
```

**Comment by BenjaminBossan** [comment]
@SP1029 Thanks for taking a look at this. Indeed, the proposed change should work. Ideally, we can have a small script for this purpose that we can run once before and once after to ensure that the outputs don't change (a regression test). It's not absolutely needed, but it could give us some peace of mind. LMK if you would like to take the task.

**Comment by SP1029** [comment]
Yes, thank you.

**Comment by SP1029** [comment]
Update: Performed a few sanity checks via existing tests.

`pytest tests/test_decoder_models.py -k test_prefix_tuning_mistral` goes through the execution path of `L789` for `DynamicCache`, and the test passes before/after the change.

Similarly, for `L796` of `EncoderDecoderCache`, the following test `pytest tests/test_encoder_decoder_models.py -k test_generate[PrefixTuningConfig` passes.

**Comment by SP1029** [comment]
> Ideally, we can have a small script for this purpose that we can run once before and once after to ensure that the outputs don't change (a regression test).

Sorry @BenjaminBossan, I got a bit confused, could you please elaborate a bit more?
Is it about adding a regression test because I'm not familiar with how, in one test, we can compare the code before and after? 
Or is it that we can separately perform two generations before and after to check?

**Comment by BenjaminBossan** [comment]
I think adding a full on regression test is overkill. I was just thinking about a small script that initializes the PEFT model and prints some output (logits or hidden states). Then we can run this script once before and once after the change and just inspect the outputs to see if they're the same. Probably we should test a few different models (decoder model, encoder-decoder model, a Gemma model) too.

**Comment by BenjaminBossan** [comment]
Closed via #2767.
